scripts/machine_translation/train_gnmt.py

```diff
@@ -195,33 +202,44 @@ def train():
     trainer = gluon.Trainer(model.collect_params(), args.optimizer, {'learning_rate': args.lr})
 
     train_data_loader, val_data_loader, test_data_loader \
-        = dataprocessor.make_dataloader(data_train, data_val, data_test, args)
+        = dataprocessor.make_dataloader(data_train, data_val, data_test, args, num_shards=len(context))
 
     best_valid_bleu = 0.0
     for epoch_id in range(args.epochs):
         log_avg_loss = 0
         log_avg_gnorm = 0
         log_wc = 0
         log_start_time = time.time()
-        for batch_id, (src_seq, tgt_seq, src_valid_length, tgt_valid_length)\
+        for batch_id, seqs \
                 in enumerate(train_data_loader):
-            # logging.info(src_seq.context) Context suddenly becomes GPU.
-            src_seq = src_seq.as_in_context(ctx)
-            tgt_seq = tgt_seq.as_in_context(ctx)
-            src_valid_length = src_valid_length.as_in_context(ctx)
-            tgt_valid_length = tgt_valid_length.as_in_context(ctx)
+
+            src_wc, tgt_wc, bs = np.sum([(shard[2].sum(), shard[3].sum(), shard[0].shape[0])
+                                         for shard in seqs], axis=0)
+
+            seqs = [[seq.as_in_context(_context) for seq in shard]
+                    for _context, shard in zip(context, seqs)]
+            src_wc = src_wc.asscalar()
+            tgt_wc = tgt_wc.asscalar()
+            LS = []
+
             with mx.autograd.record():
-                out, _ = model(src_seq, tgt_seq[:, :-1], src_valid_length, tgt_valid_length - 1)
-                loss = loss_function(out, tgt_seq[:, 1:], tgt_valid_length - 1).mean()
-                loss = loss * (tgt_seq.shape[1] - 1) / (tgt_valid_length - 1).mean()
-                loss.backward()
+                for src_seq, tgt_seq, src_valid_length, tgt_valid_length in seqs:
+                    out, _ = model(src_seq, tgt_seq[:, :-1], src_valid_length, tgt_valid_length - 1)
+                    loss = loss_function(out, tgt_seq[:, 1:], tgt_valid_length - 1).mean()
+                    loss = loss * (tgt_seq.shape[1] - 1) / (tgt_valid_length - 1).mean()
```

Comment:
The loss need to be carefully rescaled, so that the final gradient is the average of the gradients w.r.t. tokens.